### DISCUSSION TITLE
Is there a method for source fallback?

### DISCUSSION THREAD

**Discussion by kinetachien** [videojs/video.js#7235] [Q&A] [answered]
I'm working on a project that runs a live video stream that isn't always running. I need to have a fallback video but the normal way - adding a second source file - doesn't work. Is there documentation on this? Any advice?

**Comment by gkatsev** [comment, marked as answer]
There isn't really an automatic way to change sources if there is an error during playback. Your best bet is to listen to the `error` event and then set a new source.
For example:
```js
player.one('error', function() {
  player.src({
    src:: 'newsource.m3u8',
    type: 'application/x-mpegurl'
  });
});
```

On a separate note, we have a new option in Video.js 7.12 called `retryOnError`. When set, this will choose a different source if the source fails to load on initial source setting.
For example:
```html
<video-js id="vid" controls>
  <source src="source1.m3u8" type="application/x-mpegurl">
  <source src="source2.mp4" type="video/mp4">
</video-js>
```
```js
const player = videojs('vid', {
  retryOnError: true
});
```
So, with the above example, assuming that `source1.m3u8` isn't actually available, we would automatically catch that error and continue loading `source2.mp4`. Previously, we would choose `source1.m3u8` and fail playback. If `source1.m3u8` succeeds playing and fails mid-way, we will *not* switch sources.

Hope this helps!

**Reply by phloxic** [reply]
`retryOnError` seems not documented yet. Or is it?
Infering from the tests here: https://github.com/videojs/video.js/commit/22e984394292e28546e405aa6da30fc2a9e05102#diff-b62e95b3e474754b0b665ff01617843398a03815683899a9c79f141bab6a1fa7R407 it seems _not_ meant to be used in the case `canPlayType` fails, for instance to try HLS when DASH fails (iPhone) - is that correct?

**Reply by gkatsev** [reply]
It's not documented yet (we forgot) beyond what I just wrote above. iPhone shouldn't ever choose DASH if there is HLS or other playable content.

**Comment by kinetachien** [comment]
Thanks much for the replies. FWIW I ended up using readyState() along with a javascript timeout and it's working great. I had first used the error function but it took too long to check and load the fallback. 

 setTimeout(function () {
            if (player1.readyState() === 0) {
.....
